x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1148,32 +1149,37 @@ private static LogicalPlan skipPlan(UnaryPlan plan) {
     static class SkipQueryIfFoldingProjection extends OptimizerRule<LogicalPlan> {
         @Override
         protected LogicalPlan rule(LogicalPlan plan) {
-            Holder<LocalRelation> optimizedPlan = new Holder<>();
-            plan.forEachDown(Project.class, p -> {
-                List<Object> values = extractConstants(p.projections());
-                if (values.size() == p.projections().size() && (p.child() instanceof EsRelation) == false &&
-                    isNotQueryWithFromClauseAndFilterFoldedToFalse(p)) {
-                    optimizedPlan.set(new LocalRelation(p.source(), new SingletonExecutable(p.output(), values.toArray())));
-                }
-            });
+            LogicalPlan relation = plan
```

Comment:
This statement is confusing since it looks for nodes that have no children. Which currently happens to be `Relation` nodes however that's not mandatory in the future.
It's better to just introduce a `Relation` class which can be just an extension to `LeafPlan` and be the super class of all existing Relation classes (`EsRelation` and `LocalRelation`).
Then this condition becomes `plan.collect(Relation::instanceOf)`.
As an alternative that doesn't require introducing a `Relation` class simply look at the leaves and check whether they are relationship - `plan.collectLeaves()`.